src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -0,0 +1,1656 @@
+
+  return buffer;
+}
+
+// Prints one-line description of a combination of sigaction.sa_flags.
+static void print_sa_flags(outputStream* st, int flags) {
+  char buffer[0x100];
+  describe_sa_flags(flags, buffer, sizeof(buffer));
+  st->print("%s", buffer);
+}
+
+static int get_our_sigflags(int sig) {
+  assert(sig > 0 && sig < NSIG, "vm signal out of expected range");
+  return sigflags[sig];
+}
+
+static void set_our_sigflags(int sig, int flags) {
+  assert(sig > 0 && sig < NSIG, "vm signal out of expected range");
+  if (sig > 0 && sig < NSIG) {
+    sigflags[sig] = flags;
+  }
+}
+
+typedef int (*os_sigaction_t)(int, const struct sigaction *, struct sigaction *);
+
+static void SR_handler(int sig, siginfo_t* siginfo, ucontext_t* context);
+
+static void check_signal_handler(int sig) {
+  char buf[O_BUFLEN];
+  address jvmHandler = NULL;
+
+  struct sigaction act;
+  static os_sigaction_t os_sigaction = NULL;
+  if (os_sigaction == NULL) {
+    // only trust the default sigaction, in case it has been interposed
+    os_sigaction = (os_sigaction_t)dlsym(RTLD_DEFAULT, "sigaction");
+    if (os_sigaction == NULL) return;
+  }
+
+  os_sigaction(sig, (struct sigaction*)NULL, &act);
+
+
+  act.sa_flags &= SIGNIFICANT_SIGNAL_MASK;
+
+  address thisHandler = (act.sa_flags & SA_SIGINFO)
+    ? CAST_FROM_FN_PTR(address, act.sa_sigaction)
+    : CAST_FROM_FN_PTR(address, act.sa_handler);
+
+
+  switch (sig) {
+  case SIGSEGV:
+  case SIGBUS:
+  case SIGFPE:
+  case SIGPIPE:
+  case SIGILL:
+  case SIGXFSZ:
+    jvmHandler = CAST_FROM_FN_PTR(address, (sa_sigaction_t)javaSignalHandler);
+    break;
+
+  case SHUTDOWN1_SIGNAL:
+  case SHUTDOWN2_SIGNAL:
+  case SHUTDOWN3_SIGNAL:
+  case BREAK_SIGNAL:
+    jvmHandler = (address)os::user_handler();
+    break;
+
+  default:
+    if (sig == SR_signum) {
+      jvmHandler = CAST_FROM_FN_PTR(address, (sa_sigaction_t)SR_handler);
+    } else {
+      return;
+    }
+    break;
+  }
+
+  if (thisHandler != jvmHandler) {
+    tty->print("Warning: %s handler ", os::exception_name(sig, buf, O_BUFLEN));
+    tty->print("expected:%s", get_signal_handler_name(jvmHandler, buf, O_BUFLEN));
+    tty->print_cr("  found:%s", get_signal_handler_name(thisHandler, buf, O_BUFLEN));
+    // No need to check this sig any longer
+    sigaddset(&check_signal_done, sig);
+    // Running under non-interactive shell, SHUTDOWN2_SIGNAL will be reassigned SIG_IGN
+    if (sig == SHUTDOWN2_SIGNAL && !isatty(fileno(stdin))) {
+      tty->print_cr("Running in non-interactive shell, %s handler is replaced by shell",
+                    os::exception_name(sig, buf, O_BUFLEN));
+    }
+  } else if (get_our_sigflags(sig) != 0 && (int)act.sa_flags != get_our_sigflags(sig)) {
+    tty->print("Warning: %s handler flags ", os::exception_name(sig, buf, O_BUFLEN));
+    tty->print("expected:");
+    print_sa_flags(tty, get_our_sigflags(sig));
+    tty->cr();
+    tty->print("  found:");
+    print_sa_flags(tty, act.sa_flags);
+    tty->cr();
```

Comment:
Can we use unified logging at here? If do so, we should out them with single line.

Reply:
Will be addressed in https://bugs.openjdk.java.net/browse/JDK-8253742